File: packages/cli/src/commands/docs.ts
import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { defineCommand } from 'citty'
import { z } from 'zod'
import { findEntry, readAskJson } from '../io.js'
import { docsPathsFromEntry } from '../schemas.js'
import { ensureCheckout as defaultEnsureCheckout, NoCacheError } from './ensure-checkout.js'
import { findDocLikePaths } from './find-doc-paths.js'

export interface RunDocsOptions {
  spec: string
  projectDir: string
  noFetch?: boolean
  json?: boolean
}

// ── JSON output model ─────────────────────────────────────────────
// Schema for `ask docs <spec> --json`. Each candidate carries its
// source root so agents can prefer `node_modules` (faster, no clone)
// over the cached checkout, or filter on a specific origin.

export const DocsCandidateSchema = z.object({
  path: z.string(),
  root: z.enum(['node_modules', 'checkout']),
})

export const DocsModelSchema = z.object({
  spec: z.string(),
  npmPackageName: z.string().nullable(),
  checkoutDir: z.string(),
  /**
   * True iff ask.json defines a `docsPaths` override for this spec AND
   * at least one of those paths resolved to an existing file/dir.
   * When the override exists but every entry is stale, the model falls
   * back to the unfiltered walk and reports `false`.
   */
  storedOverride: z.boolean(),
  paths: z.array(DocsCandidateSchema),
})

export type DocsModel = z.infer<typeof DocsModelSchema>

export interface RunDocsDeps {
  ensureCheckout?: typeof defaultEnsureCheckout
  log?: (msg: string) => void
  error?: (msg: string) => void
  exit?: (code: number) => void
}

/**
 * Implementation of `ask docs <spec>`. Resolves the spec via the
 * shared `ensureCheckout` helper, then prints documentation candidate
 * paths from two locations (one per line):
 *
 *   1. For npm-ecosystem specs only: `node_modules/<pkg>/` if installed.
 *   2. The cached source tree (`<askHome>/github/<host>/<owner>/<repo>/<ref>/`).
 *
 * For each location, `findDocLikePaths` emits:
 *   - `dist/docs` when it exists (publish-time convention, e.g. mastra).
 *   - Any subdirectory whose basename matches `/doc/i` up to depth 4.
 *   - The root itself as a fallback when nothing above matches
 *     (README-only packages).
 *
 * Emitting only doc-like paths when they exist keeps shell substitution
 * (`rg "x" $(ask docs <spec>)`) focused on documentation instead of
 * dragging the entire source tree into the search. Callers that need
 * the checkout root itself should use `ask src`.
 */
export async function runDocs(options: RunDocsOptions, deps: RunDocsDeps = {}): Promise<void> {
  const ensureCheckout = deps.ensureCheckout ?? defaultEnsureCheckout
  const log = deps.log ?? ((msg: string) => process.stdout.write(`${msg}\n`))
  const error = deps.error ?? ((msg: string) => process.stderr.write(`${msg}\n`))
  const exit = deps.exit ?? ((code: number) => process.exit(code))
  let result
  try {
    result = await ensureCheckout({
      spec: options.spec,
      projectDir: options.projectDir,
      noFetch: options.noFetch,
    })
  }
  catch (err) {
    if (err instanceof NoCacheError) {
      error(err.message)
      exit(1)
      return
    }
    const message = err instanceof Error ? err.message : String(err)
    error(message)
    exit(1)
    return
  }

  // Collect every candidate path into `paths` so both text mode (per-
  // line `log`) and JSON mode (single blob at the end) consume the same
  // result. `emit` is the single sink — it streams to `log` in text
  // mode and just accumulates in JSON mode.
  const paths: Array<{ path: string, root: 'node_modules' | 'checkout' }> = []
  const emit = (p: string, root: 'node_modules' | 'checkout') => {
    paths.push({ path: p, root })
    if (!options.json)
      log(p)
  }

  const nmPath = result.npmPackageName
    ? path.join(options.projectDir, 'node_modules', result.npmPackageName)
    : null

  // If the spec is registered in ask.json with a persisted docsPaths
  // override, emit ONLY the stored paths (resolved against both roots
  // that `ask add` probed at selection time). Falls back to the
  // unfiltered walk when every stored path is stale — a silent
  // exact-match with no output would be worse than just printing
  // whatever does exist.
  const askJson = readAskJson(options.projectDir)
  const entry = askJson ? findEntry(askJson, options.spec) : undefined
  const stored = entry ? docsPathsFromEntry(entry) : undefined
  let storedOverride = false

  if (stored && stored.length > 0) {
    const roots: Array<{ abs: string, kind: 'node_modules' | 'checkout' }> = []
    if (nmPath && fs.existsSync(nmPath))
      roots.push({ abs: nmPath, kind: 'node_modules' })
    roots.push({ abs: result.checkoutDir, kind: 'checkout' })

    for (const rel of stored) {
      for (const { abs: root, kind } of roots) {
        // Containment guard: a malicious or buggy `docsPaths` entry
        // (`..`, absolute path) must not escape its root, otherwise
        // `ask docs` would emit arbitrary filesystem paths. Resolve
        // both sides and confirm abs is rootAbs itself or a descendant.
        const rootAbs = path.resolve(root)
        const abs = path.resolve(rootAbs, rel)
        if (abs !== rootAbs && !abs.startsWith(`${rootAbs}${path.sep}`)) {
          continue
        }
        if (fs.existsSync(abs)) {
          emit(abs, kind)
          break
        }
      }
    }

    if (paths.length > 0) {
      storedOverride = true
    }
    else {
      error(
        `ask: stored docsPaths for ${options.spec} are all stale; emitting all candidates`,
      )
      // fall through to the default walk
    }
  }

  if (!storedOverride) {
    // Walk node_modules/<pkg>/ first when the spec is an npm package
    // and the local install actually has it. Non-npm specs and missing
    // installs are silently skipped — the checkout walk below still
    // runs.
    if (nmPath && fs.existsSync(nmPath)) {
      for (const p of findDocLikePaths(nmPath))
        emit(p, 'node_modules')
    }

    // Walk the cached source tree. Always emits the root as the first
    // line, even if no /doc/i subdirs are found.
    for (const p of findDocLikePaths(result.checkoutDir))
      emit(p, 'checkout')
  }

  if (options.json) {
    const model: DocsModel = {
      spec: options.spec,
      npmPackageName: result.npmPackageName ?? null,
      checkoutDir: result.checkoutDir,
      storedOverride,
      paths,
    }
    log(JSON.stringify(DocsModelSchema.parse(model), null, 2))
  }
}

/**
 * Citty command surface for `ask docs`. Wires the citty argument
 * parser to `runDocs` and uses real stdout/stderr/process.exit. Tests
 * should call `runDocs` directly with mocked deps.
 */
export const docsCmd = defineCommand({
  meta: {
    name: 'docs',
    description: 'Print all candidate documentation paths from node_modules and the cached source tree',
  },
  args: {
    'spec': {
      type: 'positional',
      description: 'Library spec (e.g. react, npm:react@18.2.0, github:facebook/react@v18.2.0)',
      required: true,
    },
    'no-fetch': {
      type: 'boolean',
      description: 'Return cache hit only — exit 1 on cache miss',
    },
    'json': {
      type: 'boolean',
      description: 'Emit candidates as JSON matching DocsModelSchema (single blob, suppresses per-line output)',
    },
  },
  async run({ args }) {
    await runDocs({
      spec: args.spec,
      projectDir: process.cwd(),
      noFetch: Boolean(args['no-fetch']),
      json: Boolean(args.json),
    })
  },
})

File: crates/ask/src/commands/docs.rs
//! `ask docs <spec>` — print candidate documentation paths from `node_modules`
//! and the cached source tree. Rust port of `commands/docs.ts`.
//!
//! Resolves the spec via the shared `ensure_checkout` helper, then emits doc-like
//! paths (one per line, or a JSON `DocsModel` with `--json`). A persisted
//! `docsPaths` override in ask.json, when present and non-stale, restricts output
//! to those paths (resolved against both node_modules and the checkout).

use std::path::{Path, PathBuf};

use anyhow::Result;
use serde::Serialize;

use crate::commands::ensure_checkout::{
    ensure_checkout, EnsureCheckoutDeps, EnsureCheckoutOptions,
};
use crate::commands::find_doc_paths::find_doc_like_paths;
use crate::http::HttpClient;
use crate::io::{find_entry, read_ask_json};
use crate::store::assert_contained;

/// Options for [`run_docs`].
#[derive(Debug, Clone)]
pub struct RunDocsOptions {
    pub spec: String,
    pub project_dir: PathBuf,
    pub no_fetch: bool,
    pub json: bool,
}

/// Which source root a candidate path came from.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum DocsRoot {
    NodeModules,
    Checkout,
}

/// One documentation candidate path (matches `DocsCandidateSchema`).
#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct DocsCandidate {
    pub path: String,
    pub root: DocsRoot,
}

/// `ask docs <spec> --json` output (matches `DocsModelSchema`).
#[derive(Debug, Serialize, PartialEq, Eq)]
pub struct DocsModel {
    pub spec: String,
    #[serde(rename = "npmPackageName")]
    pub npm_package_name: Option<String>,
    #[serde(rename = "checkoutDir")]
    pub checkout_dir: String,
    #[serde(rename = "storedOverride")]
    pub stored_override: bool,
    pub paths: Vec<DocsCandidate>,
}

/// What [`run_docs`] produces: text to print plus any non-fatal stderr warnings
/// (e.g. all stored docsPaths were stale). The CLI prints `stdout`, writes
/// `warnings` to stderr, and exits 0 — only a hard error (NoCacheError, resolver
/// failure) surfaces as `Err` and maps to exit 1.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct DocsRun {
    pub stdout: String,
    pub warnings: Vec<String>,
}

fn candidate_at(p: &Path, root: DocsRoot) -> DocsCandidate {
    DocsCandidate {
        path: p.to_string_lossy().into_owned(),
        root,
    }
}

/// Resolve `spec` and collect documentation candidate paths.
pub fn run_docs(client: &dyn HttpClient, options: &RunDocsOptions) -> Result<DocsRun> {
    run_docs_with(client, options, &EnsureCheckoutDeps::default())
}

/// [`run_docs`] with injectable checkout deps.
pub fn run_docs_with(
    client: &dyn HttpClient,
    options: &RunDocsOptions,
    deps: &EnsureCheckoutDeps,
) -> Result<DocsRun> {
    let result = ensure_checkout(
        client,
        &EnsureCheckoutOptions {
            spec: options.spec.clone(),
            project_dir: options.project_dir.clone(),
            no_fetch: options.no_fetch,
        },
        deps,
    )?;

    let mut run = DocsRun::default();
    let mut paths: Vec<DocsCandidate> = Vec::new();

    let nm_path = result
        .npm_package_name
        .as_ref()
        .map(|pkg| options.project_dir.join("node_modules").join(pkg));

    // Persisted docsPaths override: emit ONLY the stored paths (resolved against
    // both roots that `ask add` probed). Falls back to the unfiltered walk when
    // every stored path is stale — silent empty output would be worse.
    let stored: Option<Vec<String>> =
        read_ask_json(&options.project_dir)
            .ok()
            .flatten()
            .and_then(|aj| {
                find_entry(&aj, &options.spec).and_then(|e| e.docs_paths().map(|d| d.to_vec()))
            });
    let mut stored_override = false;

    if let Some(stored) = stored.as_ref().filter(|s| !s.is_empty()) {
        // Roots in priority order: node_modules (if present) then checkout.
        let mut roots: Vec<(PathBuf, DocsRoot)> = Vec::new();
        if let Some(nm) = nm_path.as_ref().filter(|p| p.exists()) {
            roots.push((nm.clone(), DocsRoot::NodeModules));
        }
        roots.push((result.checkout_dir.clone(), DocsRoot::Checkout));

        for rel in stored {
            for (root, kind) in &roots {
                // Containment guard: a `..`/absolute docsPaths entry must not
                // escape its root. `join` mirrors path.resolve (absolute rel
                // replaces root); assert_contained rejects escapes.
                let candidate = root.join(rel);
                let Ok(abs) = assert_contained(root, &candidate) else {
                    continue;
                };
                if abs.exists() {
                    paths.push(candidate_at(&abs, *kind));
                    break;
                }
            }
        }

        if !paths.is_empty() {
            stored_override = true;
        } else {
            run.warnings.push(format!(
                "ask: stored docsPaths for {} are all stale; emitting all candidates",
                options.spec
            ));
            // fall through to the default walk
        }
    }

    if !stored_override {
        // node_modules/<pkg>/ first when the spec is an npm package and it is
        // actually installed. Non-npm / missing installs are silently skipped.
        if let Some(nm) = nm_path.as_ref().filter(|p| p.exists()) {
            for p in find_doc_like_paths(nm) {
                paths.push(candidate_at(&p, DocsRoot::NodeModules));
            }
        }
        // The cached source tree. Always emits the root as the first line even
        // when no /doc/i subdirs are found.
        for p in find_doc_like_paths(&result.checkout_dir) {
            paths.push(candidate_at(&p, DocsRoot::Checkout));
        }
    }

    run.stdout = if options.json {
        let model = DocsModel {
            spec: options.spec.clone(),
            npm_package_name: result.npm_package_name.clone(),
            checkout_dir: result.checkout_dir.to_string_lossy().into_owned(),
            stored_override,
            paths,
        };
        serde_json::to_string_pretty(&model)?
    } else {
        paths
            .iter()
            .map(|c| c.path.as_str())
            .collect::<Vec<_>>()
            .join("\n")
    };
    Ok(run)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::commands::ensure_checkout::NoCacheError;
    use crate::http::mock::MockClient;
    use crate::store::github_store_path;

    fn deps_home(home: &Path) -> EnsureCheckoutDeps<'static> {
        EnsureCheckoutDeps {
            ask_home: Some(home.to_path_buf()),
            fetcher: None,
        }
    }

    /// Pre-warm a github checkout and return (home, proj, checkout_dir).
    fn warmed() -> (tempfile::TempDir, tempfile::TempDir, PathBuf) {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        let dir = github_store_path(home.path(), "github.com", "owner", "repo", "v1.0.0").unwrap();
        std::fs::create_dir_all(&dir).unwrap();
        (home, proj, dir)
    }

    fn opts(spec: &str, proj: &Path, json: bool) -> RunDocsOptions {
        RunDocsOptions {
            spec: spec.to_string(),
            project_dir: proj.to_path_buf(),
            no_fetch: false,
            json,
        }
    }

    #[test]
    fn emits_doc_subdir_from_checkout() {
        let (home, proj, dir) = warmed();
        std::fs::create_dir_all(dir.join("docs")).unwrap();
        let run = run_docs_with(
            &MockClient::new(),
            &opts("github:owner/repo@v1.0.0", proj.path(), false),
            &deps_home(home.path()),
        )
        .unwrap();
        assert!(run.stdout.contains("docs"));
        assert!(run.warnings.is_empty());
    }

    #[test]
    fn readme_only_checkout_emits_root() {
        let (home, proj, dir) = warmed();
        std::fs::write(dir.join("README.md"), "x").unwrap();
        let run = run_docs_with(
            &MockClient::new(),
            &opts("github:owner/repo@v1.0.0", proj.path(), false),
            &deps_home(home.path()),
        )
        .unwrap();
        assert_eq!(run.stdout, dir.to_string_lossy());
    }

    #[test]
    fn json_model_has_schema_fields() {
        let (home, proj, dir) = warmed();
        std::fs::write(dir.join("README.md"), "x").unwrap();
        let run = run_docs_with(
            &MockClient::new(),
            &opts("github:owner/repo@v1.0.0", proj.path(), true),
            &deps_home(home.path()),
        )
        .unwrap();
        assert!(run.stdout.contains(r#""storedOverride": false"#));
        assert!(run.stdout.contains(r#""npmPackageName": null"#));
        assert!(run.stdout.contains(r#""root": "checkout""#));
        assert!(run.stdout.contains(r#""checkoutDir""#));
    }

    #[test]
    fn stored_docs_paths_override_restricts_output() {
        let (home, proj, dir) = warmed();
        std::fs::create_dir_all(dir.join("docs")).unwrap();
        std::fs::create_dir_all(dir.join("guides")).unwrap();
        std::fs::write(dir.join("guides/intro.md"), "x").unwrap();
        // ask.json pins docsPaths to guides/intro.md only.
        std::fs::write(
            proj.path().join("ask.json"),
            r#"{"libraries":[{"spec":"github:owner/repo@v1.0.0","docsPaths":["guides/intro.md"]}]}"#,
        )
        .unwrap();
        let run = run_docs_with(
            &MockClient::new(),
            &opts("github:owner/repo@v1.0.0", proj.path(), true),
            &deps_home(home.path()),
        )
        .unwrap();
        assert!(run.stdout.contains(r#""storedOverride": true"#));
        assert!(run.stdout.contains("guides/intro.md"));
        // Exactly one candidate — the override restricted output to the pin,
        // so the unfiltered `docs` subdir walk did not run.
        assert_eq!(run.stdout.matches(r#""path""#).count(), 1);
    }

    #[test]
    fn stale_stored_paths_warn_and_fall_through() {
        let (home, proj, dir) = warmed();
        std::fs::write(dir.join("README.md"), "x").unwrap();
        std::fs::write(
            proj.path().join("ask.json"),
            r#"{"libraries":[{"spec":"github:owner/repo@v1.0.0","docsPaths":["nonexistent/gone.md"]}]}"#,
        )
        .unwrap();
        let run = run_docs_with(
            &MockClient::new(),
            &opts("github:owner/repo@v1.0.0", proj.path(), false),
            &deps_home(home.path()),
        )
        .unwrap();
        assert_eq!(run.warnings.len(), 1);
        assert!(run.warnings[0].contains("all stale"));
        // Fell through to the unfiltered walk → root emitted.
        assert_eq!(run.stdout, dir.to_string_lossy());
    }

    #[test]
    fn containment_guard_rejects_traversal() {
        let (home, proj, dir) = warmed();
        std::fs::write(dir.join("README.md"), "x").unwrap();
        // A `..` escape must be rejected, so the override yields nothing → stale.
        std::fs::write(
            proj.path().join("ask.json"),
            r#"{"libraries":[{"spec":"github:owner/repo@v1.0.0","docsPaths":["../../../etc/hosts"]}]}"#,
        )
        .unwrap();
        let run = run_docs_with(
            &MockClient::new(),
            &opts("github:owner/repo@v1.0.0", proj.path(), false),
            &deps_home(home.path()),
        )
        .unwrap();
        assert_eq!(run.warnings.len(), 1);
        assert!(!run.stdout.contains("etc/hosts"));
    }

    #[test]
    fn no_cache_error_propagates() {
        let home = tempfile::tempdir().unwrap();
        let proj = tempfile::tempdir().unwrap();
        let err = run_docs_with(
            &MockClient::new(),
            &RunDocsOptions {
                spec: "github:owner/repo@v9.9.9".into(),
                project_dir: proj.path().to_path_buf(),
                no_fetch: true,
                json: false,
            },
            &deps_home(home.path()),
        )
        .unwrap_err();
        assert!(err.downcast_ref::<NoCacheError>().is_some());
    }
}
